tendermint/src/lite/verifier.rs

```diff
@@ -184,98 +124,120 @@ where
+    let untrusted_header = untrusted_sh.header();
+    let untrusted_commit = untrusted_sh.commit();
+    validate_signed_header_and_vals(
         untrusted_header,
-        trust_threshold,
-        trusting_period,
-        now,
-    ) {
-        Ok(_) => {
-            let untrusted_next_vals =
-                req.validator_set(untrusted_header.header().height().increment())?;
-            let untrusted_state = TS::new(untrusted_header, &untrusted_next_vals);
-            store.add(&untrusted_state)?;
-            return Ok(());
-        }
-        Err(e) => {
-            if e != Error::InsufficientVotingPower {
-                return Err(e);
+        untrusted_commit,
+        untrusted_vals,
+        untrusted_next_vals,
+    )?;
+
+    // ensure the new height is higher.
+    // if its +1, ensure the vals are correct.
+    // if its >+1, ensure we can skip to it
+    let trusted_header = trusted_state.last_header().header();
+    let trusted_height = trusted_header.height();
+    let untrusted_height = untrusted_sh.header().height();
+    match untrusted_height.cmp(&trusted_height.increment()) {
+        Ordering::Less => return Err(Error::NonIncreasingHeight),
+        Ordering::Equal => {
+            let trusted_vals_hash = trusted_header.next_validators_hash();
+            let untrusted_vals_hash = untrusted_header.validators_hash();
+            if trusted_vals_hash != untrusted_vals_hash {
+                // TODO: more specific error
+                // ie. differentiate from when next_vals.hash() doesnt
+                // match the header hash ...
+                return Err(Error::InvalidNextValidatorSet);
             }
         }
+        Ordering::Greater => {
+            let trusted_vals = trusted_state.validators();
+            verify_commit_trusting(trusted_vals, untrusted_commit, trust_threshold)?;
+        }
     }
 
-    // Here we can't trust the passed in untrusted header based on the known trusted state.
-    // Run bisection: try again with a pivot header whose height is in the
-    // middle of the trusted height and the desired height (h2.height).
-
-    let trusted_header = trusted_state.last_header();
-    let trusted_height: u64 = trusted_header.header().height().value();
-    let untrusted_height: u64 = untrusted_header.header().height().value();
-    let pivot: u64 = (trusted_height + untrusted_height) / 2;
-    let pivot_header = req.signed_header(pivot)?;
-    let pivot_vals = req.validator_set(pivot)?;
+    // verify the untrusted commit
+    verify_commit_full(untrusted_vals, untrusted_sh.commit())
+}
 
-    verify(&pivot_header, &pivot_vals)?;
+/// Attempt to update the store to the given untrusted header.
+/// Ensures our last trusted header hasn't expired yet, and that
+/// the untrusted header can be verified using only our latest trusted
+/// state from the store.
+/// This function is primarily for use by IBC handlers.
+pub fn verify_and_update_single<TS, SH, C, L, S>(
```

Comment:
Why is the side-effect of adding to the store part of this function? Wouldn't it be a better separation to decouple these two concerns.

Reply:
Good question I've gone back and forth on it. It's also currently the thing blocking the fix in the integration tests that would allow us to not expose `verify_single` because we need to mock the store there! But the idea is for this lib to provide all the functionality for maintaining a light client correctly, which includes making updates to the store. We don't want users to forget to update the store or to update it incorrectly. 

That said, forgetting is probably not really an issue, and updating incorrectly we currently don't prevent since they own the store, but if we say "we'll handle all the updates" it might help.

Also, it's partially for symmetry with the bisection function, which updates the store as it goes. We could alternatively return the set of trusted headers and remove all `store` mgmt from this lib and leave it to the user.

In any case, I think we should take this up in a separate issue ...

Reply:
Follow-up issue sounds perfect to continue this conversation and would capture my main review comment point.